### ISSUE TITLE
Questions of posed inputs (Multi-view Stereo)

### ISSUE THREAD

**Issue by haotongl** [facebookresearch/map-anything#9]
Hi, thanks for your amazing work!

When I tested your model with posed images as input, I noticed that it outputs additional camera poses.
I tried using either the predicted camera poses or the provided ground-truth poses to back-project the depth, but in both cases the results were quite inconsistent—and actually worse than when using the pose-free setting.

My questions are:

1. Why does the model output additional camera poses? Should I use the predicted poses or the GT poses when fusing the predicted depth?
2. Have you tried large-scale outdoor scenes with around 100 images in an MVS setting? I’m wondering if the multi-modal encoder has trouble generalizing.
Or do you have any suggestions?

Many thanks! I really appreciate your work, which has made feed-forward reconstruction more accessible.

```            views_example = [
                {
                    "img": (x[:, i].permute(0, 2,3,1)[0] * 255.).float().cpu().numpy().astype(np.uint8),
                    "intrinsics": kwargs["batch"]["ixt"][0, i].float().cpu().numpy(),
                    "camera_poses": affine_inverse(kwargs["batch"]["ext"][0, i]).cpu().numpy(),
                }
                for i in range(x.shape[1])
            ]
            processed_views = preprocess_inputs(views_example)
            predictions = self.model.infer(
                processed_views,                  # Any combination of input views
                memory_efficient_inference=True, # Trades off speed for more views (up to 2000 views on 140 GB)
                use_amp=True,                     # Use mixed precision inference (recommended)
                amp_dtype="bf16",                 # bf16 inference (recommended; falls back to fp16 if bf16 not supported)
                apply_mask=False,                  # Apply masking to dense geometry outputs
                mask_edges=False,                  # Remove edge artifacts by using normals and depth
                apply_confidence_mask=False,      # Filter low-confidence regions
                confidence_percentile=0,         # Remove bottom 10 percentile confidence pixels
                # Control which inputs to use/ignore
                # By default, all inputs are used when provided
                # If is_metric_scale flag is not provided, all inputs are assumed to be in metric scale
                ignore_calibration_inputs=False,
                ignore_depth_inputs=False,
                ignore_pose_inputs=False,
                ignore_depth_scale_inputs=False,
                ignore_pose_scale_inputs=False,
            )
```

<img width="2036" height="1606" alt="Image" src="https://github.com/user-attachments/assets/61c5830a-74b6-4af5-aec7-2ae71fe89a51" />

**Comment by Nik-V9** [comment]
Hi, Thanks for trying it out!

This is not expected. We have tested the MVS setting on 100-1000 views including outdoor scenes and it works great.

Response to your questions:

1. The model predicts cameras and scene geometry irrespective of the inputs provided. For getting the final prediction, you should use the predicted camera poses. However, one thing you should notice is that the predicted cameras will be very close to the GT cameras.

2. What's the camera pose convention? Have you made sure that the input poses are in OpenCV convention (XYZ - RDF) and are camera2world? The predicted camera poses should look very close to the GT cameras in the coordinate frame of first input view.

I'm happy to take a closer look if you can provide the data.

**Comment by haotongl** [comment]
@Nik-V9 Hi, thanks for your reply!

1. I noticed that the camera pose precision is much higher when I provide the camera poses compared to the pose-free setting, but the reconstruction fails (Using camera_pose to unproject depth_z).
2. I made sure that the poses are in XYZ-RDF format and follow the camera-to-world convention (just applying the inverse of the COLMAP extrinsics).
3. I’ve attached the input files, and you can run the following code to check.
[views_example.zip](https://github.com/user-attachments/files/22419764/views_example.zip)
4. By the way, does the pose input require a specific scale? Currently, the scale comes from COLMAP, which is not in metric units.
```
views_example = np.load("views_example.npy", allow_pickle=True).tolist()
processed_views = preprocess_inputs(views_example)
predictions = self.model.infer(
                processed_views,                  # Any combination of input views
                memory_efficient_inference=True, # Trades off speed for more views (up to 2000 views on 140 GB)
                use_amp=True,                     # Use mixed precision inference (recommended)
                amp_dtype="bf16",                 # bf16 inference (recommended; falls back to fp16 if bf16 not supported)
                apply_mask=False,                  # Apply masking to dense geometry outputs
                mask_edges=False,                  # Remove edge artifacts by using normals and depth
                apply_confidence_mask=False,      # Filter low-confidence regions
                confidence_percentile=0,         # Remove bottom 10 percentile confidence pixels
                # Control which inputs to use/ignore
                # By default, all inputs are used when provided
                # If is_metric_scale flag is not provided, all inputs are assumed to be in metric scale
                ignore_calibration_inputs=False,
                ignore_depth_inputs=False,
                ignore_pose_inputs=False,
                ignore_depth_scale_inputs=False,
                ignore_pose_scale_inputs=False,
)
```

Thanks again!

**Comment by Nik-V9** [comment]
Oh the scale could be the issue! Since it doesn't match what the model expects to be metric scale, it can impact performance.

When using COLMAP poses, you should pass in `is_metric_scale=False` for each view or/and use `ignore_pose_scale_inputs=True`

I can try out your data, in the mean time, let me know if this change works!

**Comment by tangjm24** [comment]
@Nik-V9  hi, I have some similar question. I use gt poses and gt intrisics as the inputs, however  the predicted poses of this method is very diference from what I give. I could ensure the script of loading camera parameters is correct(I test it on dust3r, and it works well). I WONDER WHETHER THIS METHOD SUPPORTS GT POSE FOR RECONSTRUCTION or I miss something... thanks. 

here is my code and some debugging logs.

```
image_list = load_images(images_dir)
for i,image_file in enumerate(images_files):
    img = image_list[i]['img']
    H,W = img.shape[2:]
    intrinsics = src_intrinsics[image_file.split('/')[-1]]
    extrinsics = src_extrinsics[image_file.split('/')[-1]]
    
    intrinsics[0,0] = intrinsics[0,0] * W / 1200
    intrinsics[1,1] = intrinsics[1,1] * W / 1200
    intrinsics[0,2] = W / 2
    intrinsics[1,2] = H / 2
    views_input.append({
        "img": img.squeeze().permute(1,2,0).float().cpu().numpy().astype(np.uint8),
        "intrinsics": torch.from_numpy(intrinsics).float(), 
        "camera_poses": torch.from_numpy(extrinsics).float(), 
        "is_metric_scale": torch.tensor([True], device=device),
    })


from mapanything.utils.image import preprocess_inputs

# Preprocess inputs to the expected format
processed_views = preprocess_inputs(views_input)

# Run inference with any combination of inputs
predictions = model.infer(
    processed_views,                  # Any combination of input views
    memory_efficient_inference=False, # Trades off speed for more views (up to 2000 views on 140 GB)
    use_amp=True,                     # Use mixed precision inference (recommended)
    amp_dtype="bf16",                 # bf16 inference (recommended; falls back to fp16 if bf16 not supported)
    apply_mask=True,                  # Apply masking to dense geometry outputs
    mask_edges=True,                  # Remove edge artifacts by using normals and depth
    apply_confidence_mask=False,      # Filter low-confidence regions
    confidence_percentile=10,         # Remove bottom 10 percentile confidence pixels
    # Control which inputs to use/ignore
    # By default, all inputs are used when provided
    # If is_metric_scale flag is not provided, all inputs are assumed to be in metric scale
    ignore_calibration_inputs=False,
    ignore_depth_inputs=False,
    ignore_pose_inputs=False,
    ignore_depth_scale_inputs=False,
    ignore_pose_scale_inputs=False,
    
)
```

here is some logs:

```
ipdb> intrinsics
array([[259.19198203,   0.        , 259.        ],
       [  0.        , 258.80799391, 147.        ],
       [  0.        ,   0.        ,   1.        ]])
ipdb> intrinsics_torch
tensor([[261.6930,   0.0000, 258.6030],
        [  0.0000, 260.3900, 147.1680],
        [  0.0000,   0.0000,   1.0000]], device='cuda:0')
ipdb> extrinsics
array([[-9.90348012e-01,  1.11745788e-01, -8.19981328e-02,
        -1.67862658e+00],
       [ 1.38603084e-01,  7.98447017e-01, -5.85893801e-01,
        -2.96088209e+00],
       [-4.47546848e-17, -5.91603955e-01, -8.06228727e-01,
         7.46926337e-01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])
ipdb> camera_pose_torch
tensor([[ 0.6098,  0.6216, -0.4917, -0.2524],
        [-0.0838,  0.6674,  0.7399, -0.8553],
        [ 0.7881, -0.4101,  0.4591,  2.8446],
        [ 0.0000,  0.0000,  0.0000,  1.0000]], device='cuda:0')
```

**Comment by Nik-V9** [comment]
Hi @tangjm24, there is an issue with the image normalization. `load_images` already normalizes the images with dinov2 norm and resizes them and they will get normalized again in `preprocess_input_views` (also the processing of intrinsics could be wrong if there is an cropping involved in `load_images`). 

Hence as mentioned in the README, the images should be uint8, HxWx3 images and you can leave the processing of intrinsics and other quantities to the preprocess function. When using multi-modal inputs, you shouldn’t use `load_images`. Only `preprocess_inputs` should be used.

I will fix the utility function to prevent this silent bug. For now, can you try again without using load_images function?

Also please note that the output cameras from the model are in the View 0 frame (World coordinate). So  the output view 0 (first view) camera pose should be close to identity. The input camera poses can be in any world frame, that’s fine.

**Comment by haotongl** [comment]
After I set is_metric_scale=False for each view and/or used ignore_pose_scale_inputs=True, it worked really well! Thanks again—amazing work!


<img width="2152" height="1716" alt="Image" src="https://github.com/user-attachments/assets/77212ed6-4854-453f-82d7-142a86f3cf7e" />

**Comment by tangjm24** [comment]
@Nik-V9 Thanks for your rapid reply. That's a really great work. I now could reconstruction the point cloud with good alignment with the original known poses. The only somewhat regrettable aspect is that the quality of the foreground point cloud seems to be slightly worse than the unposed cases of other methods, and I’m not sure if there are any good ways to further optimize it.

map-anything:

<img width="786" height="445" alt="Image" src="https://github.com/user-attachments/assets/576665f4-3ac2-4e0a-866d-c7c9082d750d" />

**Comment by tangjm24** [comment]
> [@Nik-V9](https://github.com/Nik-V9) Thanks for your rapid reply. That's a really great work. I now could reconstruction the point cloud with good alignment with the original known poses. The only somewhat regrettable aspect is that the quality of the foreground point cloud seems to be slightly worse than the unposed cases of other methods, and I’m not sure if there are any good ways to further optimize it.
> 
> map-anything:
> 
> <img alt="Image" width="786" height="445" src="https://private-user-images.githubusercontent.com/189321395/491911543-576665f4-3ac2-4e0a-866d-c7c9082d750d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgzNzIyNzAsIm5iZiI6MTc1ODM3MTk3MCwicGF0aCI6Ii8xODkzMjEzOTUvNDkxOTExNTQzLTU3NjY2NWY0LTNhYzItNGUwYS04NjZkLWM3YzkwODJkNzUwZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyMFQxMjM5MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYWZlOTNhZWVkYzg3OWJhOWE1OWQ2OTEzYWYyYmFkNTY4MGM1MzQ4YjAzYjI3Y2FjZGMyMTA4NGQyZjdiNzQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BkQMEvd4fsTJHpPT5IrhzZu-wpQv2_BeMqcjSDoiM0A">

compare with methods like vggt or mast3r...

<img width="690" height="505" alt="Image" src="https://github.com/user-attachments/assets/07552438-ddec-45da-985e-2ac96e2b4258" />

<img width="789" height="593" alt="Image" src="https://github.com/user-attachments/assets/14b2c258-d813-46ca-b3b2-d5d3cbdf3f64" />

map-anything has less floaters, while other methods have better alighnment...

**Comment by Nik-V9** [comment]
How many views is this and have you tried with only images as input?

**Comment by tangjm24** [comment]
> How many views is this and have you tried with only images as input?

sure, I also tried only using images. But it is even worse.  I took about 10 views for reconstruction.

**Comment by Nik-V9** [comment]
I would not expect this as Replica should be relatively easy. This appears to be a scene from Replica and believe this could be an artifact from training on [Dynamic Replica](https://dynamic-stereo.github.io/). 

I'll have to verify further. Could you share the data (the ten inputs in any format should work) and code snippet?

Also just as a note, models like DUSt3R, MASt3R, VGGT & Pi3 train on Habitat which encompasses many datasets, which also includes Replica & ReplicaCAD. So the data which you are testing with has data leak where these models have memorized the outputs. We do not use Habitat for training.

https://github.com/facebookresearch/habitat-sim/blob/main/DATASETS.md#replica-dataset

**Comment by tangjm24** [comment]
> I would not expect this as Replica should be relatively easy. This appears to be a scene from Replica and believe this could be an artifact from training on [Dynamic Replica](https://dynamic-stereo.github.io/).
> 
> I'll have to verify further. Could you share the data (the ten inputs in any format should work) and code snippet?
> 
> Also just as a note, models like DUSt3R, MASt3R, VGGT & Pi3 train on Habitat which encompasses many datasets, which also includes Replica & ReplicaCAD. So the data which you are testing with has data leak where these models have memorized the outputs. We do not use Habitat for training.
> 
> https://github.com/facebookresearch/habitat-sim/blob/main/DATASETS.md#replica-dataset

Thank you for your patient reply, and I’m sorry for getting back to you only now. My computer has been down these past few days. Attached are the data (colmap_data.zip) and the script (run.py).
You can reproduce the above results by simply running `python run.py`.


[colmap_data.zip](https://github.com/user-attachments/files/22532330/colmap_data.zip)

[run.py](https://github.com/user-attachments/files/22532272/run.py)

I look forward to your help in addressing the point cloud quality issue mentioned above.

**Comment by Nik-V9** [comment]
Please try out the new V1.1 checkpoints and let me know if you see further issues.
